src/lib/vasi-sync/src/atomic_tls_map.rs

```diff
@@ -0,0 +1,273 @@
+            table_value.write(value)
+        });
+        unsafe { Ref::new(self.values[idx].get(), &self.refcounts[idx]) }
+    }
+
+    /// Retrieve the value associated with `key`, initializing it with `init` if `key`
+    /// is not already present.
+    ///
+    /// # Safety
+    ///
+    /// No other thread must access `self` with `key` in parallel, including
+    /// having outstanding references.
+    #[inline]
+    pub unsafe fn get_or_insert_with(&self, key: NonZeroUsize, init: impl FnOnce() -> V) -> Ref<V> {
+        let val = unsafe { self.get(key) };
+        val.unwrap_or_else(|| {
+            let val = init();
+            // SAFETY: Ensured by caller
+            unsafe { self.insert(key, val) }
+        })
+    }
+
+    /// Removes the oldest value still stored for `key`, if any.
+    ///
+    /// # Safety
+    ///
+    /// *No thread* must have outstanding references to the value at `key`, or perform
+    /// parallel operations on `self` with `key`.
+    pub unsafe fn remove(&self, key: NonZeroUsize) -> Option<V> {
+        self.idx(key).map(|idx| {
+            assert_eq!(self.refcounts[idx].get(), 0);
+            let value = self.values[idx].get_mut().with(|value| {
+                let value = unsafe { &mut *value };
+                unsafe { value.assume_init_read() }
+            });
+            // Syncs with `Acquire` on insertion
+            self.keys[idx].store(None, atomic::Ordering::Release);
+            value
+        })
+    }
+}
+
+impl<const N: usize, V, H> AtomicTlsMap<N, V, H>
+where
+    H: BuildHasher + Default,
+{
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self::new_with_hasher(Default::default())
+    }
+}
+
+impl<const N: usize, V, H> Drop for AtomicTlsMap<N, V, H>
+where
+    H: BuildHasher,
+{
+    fn drop(&mut self) {
+        for idx in 0..N {
+            // Relaxed because we have exclusive access to `self`. Any other threads
+            // that have accessed the table must have synchronized with the current
+            // thread to be able to drop `self` in the first place.
+            if self.keys[idx].load(atomic::Ordering::Relaxed).is_some() {
+                self.values[idx].get_mut().with(|value| {
+                    assert_eq!(self.refcounts[idx].get(), 0);
+                    // SAFETY: We have exclusive access to `self`.
+                    let value = unsafe { &mut *value };
+                    // SAFETY: We know the value is initialized.
+                    unsafe { value.assume_init_drop() }
+                })
+            }
+        }
+    }
+}
+
+pub struct Ref<'a, V> {
+    ptr: ConstPtr<MaybeUninit<V>>,
+    refcount: &'a Cell<usize>,
+    _phantom: PhantomData<&'a V>,
+}
+
+impl<'a, V> Ref<'a, V> {
+    /// # Safety
+    ///
+    /// Current thread must be the only one to access `refcount`
+    unsafe fn new(ptr: ConstPtr<MaybeUninit<V>>, refcount: &'a Cell<usize>) -> Self {
```

Comment:
It's primarily to detect if `remove` is called on a key that's still borrowed.

Reply:
It occurred to me later that it might be a little more intuitive to make the `values` field an array of `RefCell<V>` instead of just `V`. That'd also build-in a method of doing mutable borrows, though I don't think we need that for TLS. I'm leaning slightly towards leaving it as-is, but am fine with either way.

Reply:
Added a comment to the `remove` method